database.cc

```diff
@@ -2268,6 +2268,70 @@ future<> database::clear_snapshot(sstring tag, std::vector<sstring> keyspace_nam
     });
 }
 
+future<std::unordered_map<sstring, std::unordered_map<std::pair<sstring, sstring>, db::snapshot_ctl::snapshot_details, utils::tuple_hash>>> database::get_snapshot_details() {
+    co_return co_await seastar::async([this] () {
+        std::vector<sstring> data_dirs = _cfg.data_file_directories();
+        auto dirs_only_entries_ptr =
+            make_lw_shared<lister::dir_entry_types>(lister::dir_entry_types{directory_entry_type::directory});
+        std::unordered_map<sstring, std::unordered_map<std::pair<sstring, sstring>, snapshot_ctl::snapshot_details, utils::tuple_hash>> details;
+
+        for (auto& datadir : data_dirs) {
+            lister::scan_dir(datadir, *dirs_only_entries_ptr, [this, dirs_only_entries_ptr, &details] (fs::path parent_dir, directory_entry de) mutable {
+                // KS directory
+                sstring ks_name = de.name;
+
+                return lister::scan_dir(parent_dir / de.name, *dirs_only_entries_ptr, [this, dirs_only_entries_ptr, &details, ks_name = std::move(ks_name)] (fs::path parent_dir, directory_entry de) mutable {
+                    // CF directory
+                    sstring cf_name = de.name.substr(0, de.name.find("-"));
+                    fs::path cf_dir = parent_dir / de.name;
+
+                    return lister::scan_dir(parent_dir / de.name, *dirs_only_entries_ptr, [this, dirs_only_entries_ptr, &details, ks_name, cf_name = std::move(cf_name), cf_dir = std::move(cf_dir)] (fs::path parent_dir, directory_entry de) mutable {
+                        // "snapshots" directory
```

Comment:
what about other subdirectories like "staging", "upload", etc?
please see my `compaction: quarantine invalid sstables` series on scylladb-dev the mailing list
as I'm adding yet another subdirectory.
and in particular, `[PATCH v4 01/11] sstables: define symbolic names for table subdirectories`
that introduces `sstables::snapshots_dir` that you could use for listing only the snapshots subdirectory.

Reply:
Snapshots can only appear in the snapshots directory, can't they? So any other directories are not relevant to us.

As for the `sstables::snapshots_dir` it does look like a nice thing to use. The patch did not go in yet, did it? Can I use it before it's introduced (e.g. add the `constexpr ... snapshots_dir = ...` myself)?

Reply:
no, it didn't go in yet
Feel free to cherry pick the needed patch from my branch:

https://github.com/bhalevy/scylla/commit/2804fc4ee07b7eb9f20eea3a10c209b360a4bb92

Reply:
The whole branch is https://github.com/bhalevy/scylla/commits/quarantine-invalid-sstables